### DISCUSSION TITLE
How do you get `app_scoped_privkey` on the frontend for server side verification?

### DISCUSSION THREAD

**Discussion by JSeam2** [Web3Auth/Web3Auth#433] [Help]
I need some clarifications regarding the `app_scoped_privkey` as this seems like a crucial detail you need to authenticate the jwt tokens obtained from the `idToken`

I am following the tutorial found here:
https://web3auth.io/docs/server-side-verification/social-login-users

I know how to obtain the `idToken`. This is possible using something like `const user = await web3AuthInstance.getUserInfo();`

However, there seems to be missing documentation regarding the `app_scoped_privkey` within the web3auth docs. 

The documentation states
```
App scoped private key will be obtained from the frontend once the user is logged in.
```
and then shows the following snippets
```javascript
// Incase of ed25519 curve
import { getED25519Key } from "@toruslabs/openlogin-ed25519";
const app_scoped_privkey = "app scoped private key";
const ed25519Key = getED25519Key(Buffer.from(app_scoped_privkey.padStart(64, "0"), "hex"));
const app_pub_key = ed25519Key.pk.toString("hex");

// Incase of secp256k1 curve
import { getPublicCompressed } from "@toruslabs/eccrypto";
const app_scoped_privkey = "app scoped private key";
const app_pub_key = getPublicCompressed(Buffer.from(app_scoped_privkey.padStart(64, "0"), "hex")).toString("hex");
```

There seems to be missing instructions regarding how `app_scoped_privkey` can be obtain atm. Would need some help on this

**Comment by JSeam2** [comment]
Ok I think I have solved this was observing this tutorial found here [https://github.com/Web3Auth/examples/blob/main/web-core-sdk/server-side-verification/ssv-via-social-nextjs-core-example/pages/App.tsx](https://github.com/Web3Auth/examples/blob/main/web-core-sdk/server-side-verification/ssv-via-social-nextjs-core-example/pages/App.tsx)

Gist of how to get the private key, you will need to call the `web3auth.provider` to obtain the private key which then allows you to obtain the public key using the `@toruslabs/eccrypto` library using the `getPublicCompressed()` function. 

This snippet or link to the sample implementation should be included in the Server Side Verification docs here [https://web3auth.io/docs/server-side-verification/social-login-users](https://web3auth.io/docs/server-side-verification/social-login-users). It is an important missing step
```javascript
  const getUserInfo = async () => {
    if (!web3auth) {
      uiConsole('web3auth not initialized yet')
      return
    }
    const user = await web3auth.getUserInfo()

    const privKey = await web3auth.provider?.request({
      method: 'eth_private_key',
    })
    const pubkey = getPublicCompressed(Buffer.from(privKey, 'hex')).toString(
      'hex',
    )

    // Validate idToken with server
    const res = await fetch('/api/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: 'Bearer ' + user.idToken,
      },
      body: JSON.stringify({appPubKey: pubkey}),
    })
    if (res.status === 200) {
      console.log('JWT Verification Successful')
      uiConsole(user)
    } else {
      console.log('JWT Verification Failed')
      uiConsole('JWT Verification Failed')
    }
  }
```

**Reply by dangdennis** [reply]
Why do we need the appkey for server side verification? We already have your public jwks to verify the jwt (idToken) as a whole. If we don't verify the app scoped public key, what attack vector are we exposing ourselves to?

**Reply by JSeam2** [reply]
It seems to be something that’s recommended on the docs atm.

I think some kind of signing to validate the signature would make more sense tbh. Now that you’ve mentioned, the flow used does give out the same public key. Would like the web3auth team to clarify on this.

In terms of attack vectors the possibility of calling the priv_key into the memory of some browser app could be the priv_key being stolen by some back door on either the browser or host device.

Another attack vector of not verifying some signature signed using the priv_key is the possibility of a man in the middle where someone sniffs out the public_key stores in the idToken and pretends to be that identity.
